### ISSUE TITLE
Installing the collector on ECS Fargate

### ISSUE THREAD

**Issue by sj26** [pganalyze/pganalyze-docs#2]
We recently switched our pganalyze collector to run in [ECS](https://aws.amazon.com/ecs/) as a [Fargate](https://aws.amazon.com/fargate/) task, which is neat because it means no servers are required and lowers the operational burden. Here's a rough version of what we did.

(I was going to PR some test, but all I found in the installing collector docs page was a custom element, `<docs-collector-install></docs-collector-install>`, so here are some notes instead.)

We have [a task definition](https://docs.aws.amazon.com/AmazonECS/latest/userguide/task_definitions.html) with a container definition like:

```json
    {
      "name": "pganalyze",
      "image": "quay.io/pganalyze/collector:stable",
      "essential": true,
      "environment": [
        {"name": "FILTER_QUERY_SAMPLE", "value": "all"},
        {"name": "FILTER_LOG_SECRET", "value": "all"},

        {"name": "DB_HOST", "value": "..."},
        {"name": "DB_PORT", "value": "..."},
        {"name": "DB_SSLMODE", "value": "require"},
        {"name": "DB_SSLROOTCERT", "value": "rds-ca-2015-root"},
        {"name": "DB_USER", "value": "pganalyze"},
        {"name": "DB_NAME", "value": "our_database"},

        {"name": "AWS_REGION", "value": "us-east-1"},
        {"name": "AWS_INSTANCE_ID", "value": "our_instance"}
      ],
      "secrets": [
        {"name": "PGA_API_KEY", "valueFrom": "/pganalyze/PGA_API_KEY"},
        {"name": "DB_PASSWORD", "valueFrom": "/pganalyze/DB_PASSWORD"}
      ],
      "logConfiguration": {
        "logDriver": "awslogs",
        "options": {
          "awslogs-group": "/ecs/pganalyze",
          "awslogs-region": "us-east-1",
          "awslogs-stream-prefix": "pganalyze"
        }
      },
      "readonlyRootFilesystem": false,
      "mountPoints": []
    }
```

Notably the AWS_INSTANCE_ID instance detection from the DB_HOST doesn't seem to work when using env, so I had to add that and the AWS_REGION to get things working. I had to code dive a little to find these environment variable names, too — it'd be nice to have a reference in the docs. Otherwise it's been great!

We actually have our own ECS Repository which we cross-publish the images into for security control and availability reasons, and because I [patched the collector a little](https://github.com/pganalyze/collector/pull/35), but using the image directly from quay.io will work too (once that patch PR is merged).

We stored our secure parameters `PGA_API_KEY` and `DB_PASSWORD` as [SSM Parameters](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html) with SecureString values like:

```
aws ssm put-parameter --name /pganalyze/PGA_API_KEY --type SecureString --value "$PGA_API_KEY"

aws ssm put-parameter --name /pganalyze/DB_PASSWORD --type SecureString --value "$DB_PASSWORD
``` 

We use this with an IAM role like this, although this has been generalised and the security scope widened so it'll work on copy/paste — note that you'll probably want to specify the kms key id used to encrypt the above ssm parameters instead of the wildcard:

```
aws iam create-role --role-name pganalyzeTaskRole \
  --description "pganalyze collector ecs tasks" \
  --assume-role-policy-document '{"Version": "2012-10-17", "Statement": [{"Effect": "Allow", "Principal": {"Service": "ecs-tasks.amazonaws.com"}, "Action": "sts:AssumeRole"}]}'

aws iam attach-role-policy --role-name pganalyzeTaskRole \
  --policy-arn arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy

aws iam put-role-policy --role-name pganalyzeTaskRole \
  --policy-name DescribeEC2Metadata \
  --policy-document '{"Statement":[{"Action": ["ec2:DescribeAccountAttributes", "ec2:DescribeAvailabilityZones", "ec2:DescribeSecurityGroups", "ec2:DescribeVpcs"], "Effect": "Allow", "Resource": "*"}]}'

aws iam put-role-policy --role-name pganalyzeTaskRole \
  --policy-name GetCloudWatchMetricsStatistics \
  --policy-document '{"Statement":[{"Action": "cloudwatch:GetMetricStatistics", "Effect": "Allow", "Resource": "*"}]}'

aws iam put-role-policy --role-name pganalyzeTaskRole \
  --policy-name GetCloudWatchLogs \
  --policy-document '{"Statement":[{"Action": ["logs:DescribeLogStreams", "logs:GetLogEvents"], "Effect": "Allow", "Resource": "*"}]}'

aws iam put-role-policy --role-name pganalyzeTaskRole \
  --policy-name DescribeRdsDatabases \
  --policy-document '{"Statement":[{"Action": ["rds:Describe*", "rds:ListTagsForResource"], "Effect": "Allow", "Resource": "*"}]}'

aws iam put-role-policy --role-name pganalyzeTaskRole \
  --policy-name GetRdsLogs \
  --policy-document '{"Statement":[{"Action": "rds:DownloadDBLogFilePortion", "Effect": "Allow", "Resource": "*"}]}'

aws iam put-role-policy --role-name pganalyzeTaskRole \
  --policy-name GetSSMPganalyzeParameters \
  --policy-document '{"Statement":[{"Action": ["ssm:GetParameter", "ssm:GetParameters", "ssm:GetParametersByPath"], "Effect": "Allow", "Resource": "arn:aws:ssm:*:*:parameter/pganalyze/*"}, {"Action": "kms:Decrypt", "Effect": "Allow", "Resource": "arn:aws:kms:*:*:*"}]}'
```

And the task definition is run as a single persistent task using an [ECS Service](https://docs.aws.amazon.com/AmazonECS/latest/userguide/ecs_services.html). Make sure to use `--launch-type FARGATE --platform-version 1.3.0` so that the SSM secrets work.

Works great!

**Comment by lfittl** [comment]
@sj26 Excellent, thanks for sharing this detailed information!

I'll see that we work this into the documentation - will update this issue once its in.

**Comment by lfittl** [comment]
This has been added to the in-app documentation as of a few weeks ago - thanks again for the contribution!
